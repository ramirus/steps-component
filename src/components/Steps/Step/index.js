import React from 'react'
import styled from 'styled-components';

// template props
const tp = {
  activeColor: '#5025D6',
  secondColor: '#DBDBDB',
  iconSize: '40px',
}

const Step = (props) => {
  const ButtonContainer = styled.div`
    position: relative;

    &:after {
      content: '';
      height: 8px;
      width: 9999px;
      background-color: ${props.checked ? tp.activeColor : tp.secondColor};
      display: block;
      position: absolute;
      top: 16px;
      left: 50%;
      z-index: -1;
    }
  `

  const Container = styled.div`
    flex: 1;

    &:last-child {
      flex: none;
      ${ButtonContainer}: {
        &:after {
          background: #fff;
        }
      }
    }
  `

  const Element = styled.div`
    display: inline-block;
    text-align: center;
  `

  const Label = styled.label`
    display: block;
    padding-bottom: 15px;
    font-size: 18px;
    font-weight: bold;
    color: ${props.checked || props.current ? tp.activeColor : tp.secondColor};
  `
  Label.displayName = 'Label'

  const Button = styled.input`
    width: 40px;
    height: 40px;
    border: 10px solid ${props.checked || props.current ? tp.activeColor : tp.secondColor};
    border-radius: 50%;
    padding: 0;
    margin: 0;
    background-color: #fff;
    cursor: ${props.active ? 'pointer' : 'default'};
  `
  Button.displayName = 'Button'

  const handleClick = (e) => {
    e.preventDefault();
    props.handleClick(props.index);
  }

  return (
    <Container>
      <Element>
        <Label htmlFor={props.title}>{props.title}</Label>
        <ButtonContainer>
          <Button
            id={props.title}
            type="button"
            onClick={handleClick}
            disabled={!props.active}
          />
        </ButtonContainer>
      </Element>
    </Container>
  )
}

export default Step
