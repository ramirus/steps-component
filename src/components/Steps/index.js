import React from 'react'
import Step from './Step'

const Steps = (props) => {

  const isActive = (index) => Math.abs(index - props.current)

  const handleClick = (index) => {
    if (isActive(index)) {
      this.props.handleClick(index)
    }
  }

  const renderChildren = () => {
    const { current, steps } = props;

    const stepLengtn = steps.length;
    if (stepLengtn < 2) {
      console.warn("In Steps component must be more than 2 and max 5 steps");
      return null;
    }
    if (stepLengtn > 2) {
      console.warn("In Steps component must be max 5 steps");
    }

    const result = [];
    
    for (let i = 0; i < stepLengtn && i < 5; i++) {
      result.push(
        <Step
          key={steps[i].label}
          title={steps[i].label}
          handleClick={handleClick}
          current={current === i}
          active={isActive(i)}
          checked={i < current}
        />
      )
    }

    return result;
  }
  

  return (
    <div>
      { renderChildren() }
    </div>
  )
}

export default Steps
