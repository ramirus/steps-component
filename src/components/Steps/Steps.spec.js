import React from 'react'
import { shallow } from 'enzyme'
import Steps from './index'

describe('Steps container', () => {
  const props = {
    steps: [],
    current: 0,
  }

  describe('Steps container initial', () => {
    const steps = shallow(<Steps {...props} />)

    it('renders properly', () => {
      expect(steps).toMatchSnapshot()
    })
  
    it('not render <Step /> if "steps" prop is empty', () => {
      expect(steps.find('Step')).toHaveLength(0)
    })
  })

  describe('Steps container has one child', () => {
    const nextProps = {
      ...props,
      steps: [
        {
          label: 'Design',
        },
      ]
    }

    const steps = shallow(<Steps {...nextProps} />)

    it('not render <Step /> if "steps" prop length is 1', () => {
      expect(steps.find('Step')).toHaveLength(0)
    })
  })

  describe('Steps container render <Step />', () => {
    const nextProps = {
      ...props,
      steps: [
        {
          label: 'Design',
        },
        {
          label: 'Build',
        },
        {
          label: 'Launch',
        },
      ]
    }

    const steps = shallow(<Steps {...nextProps} />)

    it('render <Step /> components', () => {
      expect(steps.find('Step')).toHaveLength(nextProps.steps.length)
    })
  })

  describe('Max steps props length in  Steps container', () => {
    const nextProps = {
      ...props,
      steps: [
        {
          label: 'Design',
        },
        {
          label: 'Build',
        },
        {
          label: 'Launch',
        },
        {
          label: 'Launch 1',
        },
        {
          label: 'Launch 2',
        },
        {
          label: 'Launch 3',
        },
      ]
    }

    const steps = shallow(<Steps {...nextProps} />)

    it('If steps props length > 5 Steps container render only 5 <Step /> components', () => {
      expect(steps.find('Step')).toHaveLength(5)
    })
  })
})
