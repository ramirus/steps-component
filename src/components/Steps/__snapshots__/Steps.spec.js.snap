// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Steps container Steps container initial renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Steps
    current={0}
    steps={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": null,
    },
    "ref": null,
    "rendered": null,
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-bdVaJa",
        "isStatic": true,
        "rules": Array [
          "
  display: flex;
  padding: 10px;
",
        ],
      },
      "displayName": "styled.div",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "sc-bdVaJa",
      "target": "div",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": null,
      },
      "ref": null,
      "rendered": null,
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-bdVaJa",
          "isStatic": true,
          "rules": Array [
            "
  display: flex;
  padding: 10px;
",
          ],
        },
        "displayName": "styled.div",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-bdVaJa",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
